[inductor][ck] kBatch filtering with gen_ops

coconutruben opened the pull request:
Summary:
# Why

not all choices of kBatch are valid and will lead to a runtime error (when CK checks the validity of the args)

https://github.com/ROCm/composable_kernel/blob/c9bcfd755ed4d2102d76a6f545ac6e9a030d7d8e/include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_v3_multi_d.hpp#L1020

# What

- move kBatch inside the gen_ops to have more control over it, and be able to filter it
- expand filtering based on the cpp logic
- refactor the padding checks to be more readable

Test Plan:
```
buck2 run -c fbcode.re_gpu_tests=False mode/opt-amd-gpu  fbcode//deeplearning/aot_inductor/benchmark/sampling:test_gemm_autotune_benchmark_AMD_block_0
```

with

kBatch = 128: some filering
kBatch = 1: no filering
kBatch = 1738: all options filtered out

Reviewed By: henrylhtsang

Differential Revision: D70211442




cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

ColinPeppler reviewed (approved):
i know a bit unpleasant but make sure to fix linter

tenpercent commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -451,12 +486,81 @@ def filter_op(self, op: "CKGemmOperation"):
+                num_k_loop = ak0 // (op.k_per_block // op.a_k1)
+                if num_k_loop <= prefetch_stages:
+                    log.debug(
+                        f"Op {op.name()} is not compatible due to invalid number of pipeline prefetch stages. "
```
I remember some loggers complain about f-strings in the log statements

coconutruben replied:
you're right! I'll fix that in a subsequent pr :)

tenpercent reviewed (approved):
one small comment, otherwise looks good, thanks!

coconutruben committed:
[inductor][ck] kBatch filtering with gen_ops (#148004)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/148004

# Why

not all choices of kBatch are valid and will lead to a runtime error (when CK checks the validity of the args)

https://github.com/ROCm/composable_kernel/blob/c9bcfd755ed4d2102d76a6f545ac6e9a030d7d8e/include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_cshuffle_v3_multi_d.hpp#L1020

# What

- move kBatch inside the gen_ops to have more control over it, and be able to filter it
- expand filtering based on the cpp logic
- refactor the padding checks to be more readable

Test Plan:
```
buck2 run -c fbcode.re_gpu_tests=False mode/opt-amd-gpu  fbcode//deeplearning/aot_inductor/benchmark/sampling:test_gemm_autotune_benchmark_AMD_block_0
```

with

kBatch = 128: some filering
kBatch = 1: no filering
kBatch = 1738: all options filtered out

Reviewed By: ColinPeppler, henrylhtsang

Differential Revision: D70211442
